src/cm/communication_manager.rs

```diff
@@ -56,55 +229,356 @@ impl EventChannel {
+        let mut user_data = self.user_context.lock().unwrap();
+        *user_data = Some(Arc::new(ctx));
+    }
+
+    pub fn get_context<C: Any + Send + Sync>(&self) -> Option<Arc<C>> {
+        let user_data = self.user_context.lock().unwrap();
+        let arc_any = user_data.as_ref()?.clone();
+        arc_any.downcast::<C>().ok()
+    }
+
+    pub fn port(&self) -> u8 {
+        let cm_id = self.cm_id;
+
+        unsafe { cm_id.as_ref().port_num }
+    }
+
+    pub fn bind_addr(&self, addr: SocketAddr) -> Result<(), String> {
+        let cm_id = self.cm_id;
+        let ret = unsafe { rdma_bind_addr(cm_id.as_ptr(), OsSocketAddr::from(addr).as_mut_ptr()) };
+
+        if ret < 0 {
+            return Err(format!("Failed to bind addr {addr:?}, {}", io::Error::last_os_error()));
+        }
+
+        Ok(())
+    }
+
+    pub fn resolve_addr(
+        &self, src_addr: Option<SocketAddr>, dst_addr: SocketAddr, timeout: Duration,
+    ) -> Result<(), String> {
+        let cm_id = self.cm_id;
+        let timeout_ms: i32 = timeout.as_millis().try_into().unwrap();
+
+        let ret = unsafe {
+            rdma_resolve_addr(
+                cm_id.as_ptr(),
+                match src_addr {
+                    Some(addr) => OsSocketAddr::from(addr).as_mut_ptr(),
+                    None => null_mut(),
+                },
+                OsSocketAddr::from(dst_addr).as_mut_ptr(),
+                timeout_ms,
+            )
+        };
 
         if ret < 0 {
-            return Err(format!("Failed to bind addr {addr:?}, returned {ret}"));
+            return Err(format!("Failed to resolve address {ret}"));
         }
 
         Ok(())
     }
 
-    pub fn listen(&mut self, backlog: i32) -> Result<(), String> {
-        let ret = unsafe { rdma_listen(self.cm_id.as_mut(), backlog) };
+    pub fn resolve_route(&self, timeout: Duration) -> Result<(), String> {
+        let cm_id = self.cm_id;
+        let timeout_ms: i32 = timeout.as_millis().try_into().unwrap();
+
+        let ret = unsafe { rdma_resolve_route(cm_id.as_ptr(), timeout_ms) };
+
+        if ret < 0 {
+            return Err(format!("Failed to resolve route {ret}"));
+        }
+
+        Ok(())
+    }
+
+    pub fn listen(&self, backlog: i32) -> Result<(), String> {
+        let cm_id = self.cm_id;
+        let ret = unsafe { rdma_listen(cm_id.as_ptr(), backlog) };
 
         if ret < 0 {
             return Err(format!("Failed to listen {ret}"));
         }
 
         Ok(())
     }
+
+    pub fn get_device_context(&self) -> Option<Arc<DeviceContext>> {
+        let cm_id = self.cm_id;
+
+        unsafe {
+            let mut guard = DEVICE_LISTS.lock().unwrap();
+            let device_ctx = guard
+                .entry((*cm_id.as_ptr()).verbs as usize)
+                .or_insert(Arc::new(DeviceContext {
+                    context: (*cm_id.as_ptr()).verbs,
+                }));
+
+            Some(device_ctx.clone())
```

Comment:
It seems that `get_device_context()` will always return a valid `DeviceContext` object. Maybe we can return `Arc<DeviceContext>` directly?

Reply:
Hmm you just remind me that newly created cm_ids that haven't bound any addresses would have no device context, that would lead to return None, though the logic here is still problematic